h2o-persist-hdfs/src/main/java/water/persist/PersistHdfs.java

```diff
@@ -315,34 +326,34 @@ private static void addFolder(FileSystem fs, Path p, ArrayList<String> keys, Arr
       futures.blockForPending();
     }
   }
+  
+  private static boolean isDuplicatedRootFolderPresent(final FileStatus fileStatus, final Path referencePath){
+    return fileStatus.getPath().getName().equals(referencePath.getName());
+  }
 
   @Override
   public Key uriToKey(URI uri) throws IOException {
     assert "hdfs".equals(uri.getScheme()) || "s3".equals(uri.getScheme())
             || "s3n".equals(uri.getScheme()) || "s3a".equals(uri.getScheme()) : "Expected hdfs, s3 s3n, or s3a scheme, but uri is " + uri;
 
+      final String uriWithCredentials = encodeCredentialsIfPresent(uri.toString());
+      if(uriWithCredentials != null) {
+        uri = URI.create(uriWithCredentials);
+      }
+
     FileSystem fs = FileSystem.get(uri, PersistHdfs.CONF);
     FileStatus[] fstatus = fs.listStatus(new Path(uri));
     assert fstatus.length == 1 : "Expected uri to single file, but uri is " + uri;
-
-    return HDFSFileVec.make(fstatus[0].getPath().toString(), fstatus[0].getLen());
+    // The FileSystem API may return the path with decoded credentials - thus it is required to re-encode.
+    return HDFSFileVec.make(encodeCredentialsIfPresent(fstatus[0].getPath().toString()), fstatus[0].getLen());
   }
-
-  public static FileSystem getFS(String path) throws IOException {
-    try {
-      return getFS(new URI(path));
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(e);
-    }
-  }
-  public static FileSystem getFS(URI uri) throws IOException {
-    return FileSystem.get(uri, PersistHdfs.CONF);
-  }
-
+  
+  
   // Is there a bucket name without a trailing "/" ?
+  private static final Pattern BUCKET_ONLY_PATTERN = Pattern.compile("s3n://.@{0,1}[^/]*");
```

Comment:
Historically limited to s3n. These APIs hide so many traps I'd not be surprised if s3a and s3n behaved differently and this method was really meant to check just s3n.